extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -0,0 +1,423 @@
+      final String topic = getKafkaTopic();
+      LOG.debug("About to listen to topic [%s] with group.id [%s]", topic, factoryId);
+      final Map<String, String> map = cacheManager.getCacheMap(factoryId);
+      mapRef.set(map);
+      // Enable publish-subscribe
+      kafkaProperties.setProperty("auto.offset.reset", "smallest");
+
+      final CountDownLatch startingReads = new CountDownLatch(1);
+
+      final ListenableFuture<?> future = executorService.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              while (!executorService.isShutdown() && !Thread.currentThread().isInterrupted()) {
+                final ConsumerConnector consumerConnector = buildConnector(kafkaProperties);
+                try {
+                  final List<KafkaStream<String, String>> streams = consumerConnector.createMessageStreamsByFilter(
+                      new Whitelist(Pattern.quote(topic)), 1, DEFAULT_STRING_DECODER, DEFAULT_STRING_DECODER
+                  );
+
+                  if (streams == null || streams.isEmpty()) {
+                    throw new IAE("Topic [%s] had no streams", topic);
+                  }
+                  if (streams.size() > 1) {
+                    throw new ISE("Topic [%s] has %d streams! expected 1", topic, streams.size());
+                  }
+                  final KafkaStream<String, String> kafkaStream = streams.get(0);
+
+                  startingReads.countDown();
+
+                  for (final MessageAndMetadata<String, String> messageAndMetadata : kafkaStream) {
+                    final String key = messageAndMetadata.key();
+                    final String message = messageAndMetadata.message();
+                    if (key == null || message == null) {
+                      LOG.error("Bad key/message from topic [%s]: [%s]", topic, messageAndMetadata);
+                      continue;
+                    }
+                    doubleEventCount.incrementAndGet();
+                    map.put(key, message);
+                    doubleEventCount.incrementAndGet();
+                    LOG.trace("Placed key[%s] val[%s]", key, message);
+                  }
+                }
+                catch (Exception e) {
+                  LOG.error(e, "Error reading stream for topic [%s]", topic);
+                }
+                finally {
+                  consumerConnector.shutdown();
+                }
+              }
+            }
+          }
+      );
+      Futures.addCallback(
+          future, new FutureCallback<Object>()
+          {
+            @Override
+            public void onSuccess(Object result)
+            {
+              LOG.debug("Success listening to [%s]", topic);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              if (t instanceof CancellationException) {
+                LOG.debug("Topic [%s] cancelled", topic);
```

Comment:
During normal shutdown the thread is interrupted and has the future's cancel flag set to true. I can put it as info if you want but warning is not correct.